src/ModelContextProtocol/Server/McpServer.cs

```diff
@@ -79,55 +113,60 @@ public McpServer(ITransport transport, McpServerOptions options, ILoggerFactory?
 
-    /// <inheritdoc />
-    public async Task StartAsync(CancellationToken cancellationToken = default)
+    public async Task AcceptSessionAsync(CancellationToken cancellationToken = default)
     {
-        if (_isInitializing)
-        {
-            _logger.ServerAlreadyInitializing(EndpointName);
-            throw new InvalidOperationException("Server is already initializing");
-        }
-        _isInitializing = true;
-
-        if (IsInitialized)
-        {
-            _logger.ServerAlreadyInitializing(EndpointName);
-            return;
-        }
+        // Below is effectively an assertion. The McpServerFactory should only use this with the IServerTransport constructor.
+        Throw.IfNull(_serverTransport);
 
         try
         {
-            CancellationTokenSource = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
+            _sessionTransport = await _serverTransport.AcceptAsync(cancellationToken).ConfigureAwait(false);
 
-            if (_serverTransport is not null)
+            if (_sessionTransport is null)
             {
-                // Start listening for messages
-                await _serverTransport.StartListeningAsync(CancellationTokenSource.Token).ConfigureAwait(false);
+                throw new McpServerException("The server transport closed before a client started a new session.");
             }
 
-            // Start processing messages
-            MessageProcessingTask = ProcessMessagesAsync(CancellationTokenSource.Token);
-
-            // Unlike McpClient, we're not done initializing until we've received a message from the client, so we don't set IsInitialized here
+            InitializeSession(_sessionTransport);
         }
         catch (Exception e)
         {
             _logger.ServerInitializationError(EndpointName, e);
-            await CleanupAsync().ConfigureAwait(false);
             throw;
         }
     }
 
-    protected override Task CleanupAsync()
+    /// <inheritdoc />
+    public async Task RunAsync(CancellationToken cancellationToken = default)
+    {
+        try
+        {
+            // Start processing messages
+            StartSession(fullSessionCancellationToken: cancellationToken);
+            await MessageProcessingTask.ConfigureAwait(false);
+        }
+        finally
+        {
+            await DisposeAsync().ConfigureAwait(false);
+        }
+    }
+
+    public override async ValueTask DisposeUnsynchronizedAsync()
     {
         if (ServerOptions.Capabilities?.Tools?.ToolCollection is { } tools)
         {
             tools.Changed -= _toolsChangedDelegate;
         }
-        return base.CleanupAsync();
+
+        await base.DisposeUnsynchronizedAsync().ConfigureAwait(false);
+
+        if (_serverTransport is not null && _sessionTransport is not null)
+        {
+            // We created the _sessionTransport from the _serverTransport, so we own it.
+            await _sessionTransport.DisposeAsync().ConfigureAwait(false);
```

Comment:
Do we want to do this even if base.DisposeUnsynchronizedAsync above throws?

Reply:
Good call. I'll dispose the session transport in a finally both here and in the client.